tests/schema/tests.py

```diff
@@ -275,6 +275,46 @@ def test_inline_fk(self):
             if sql.startswith('ALTER TABLE') and 'ADD CONSTRAINT' in sql
         ])
 
+    @skipUnlessDBFeature('can_create_inline_fk')
+    def test_add_inline_fk_update_data(self):
+        with connection.schema_editor() as editor:
+            editor.create_model(Node)
+        # Add an inline foreign key and update data in the same transaction.
+        new_field = ForeignKey(Node, CASCADE, related_name='new_fk', null=True)
+        new_field.set_attributes_from_name('new_parent_fk')
+        parent = Node.objects.create()
+        with connection.schema_editor() as editor:
+            editor.add_field(Node, new_field)
+            editor.execute('UPDATE schema_node SET new_parent_fk_id = %s;', [parent.pk])
+        self.assertIn('new_parent_fk_id', self.get_indexes(Node._meta.db_table))
+
+    @skipUnlessDBFeature(
+        'can_create_inline_fk',
+        'allows_multiple_constraints_on_same_fields',
+    )
+    def test_add_inline_fk_index_update_data(self):
+        with connection.schema_editor() as editor:
+            editor.create_model(Node)
+        # Add an inline foreign key, update data, and an index manually in the
+        # same transaction.
+        new_field = ForeignKey(Node, CASCADE, related_name='new_fk', null=True)
+        new_field.set_attributes_from_name('new_parent_fk')
+        parent = Node.objects.create()
+        with connection.schema_editor() as editor:
+            editor.add_field(Node, new_field)
+            editor.execute('UPDATE schema_node SET new_parent_fk_id = %s;', [parent.pk])
+            editor.execute(
+                editor.sql_create_index % {
+                    'table': editor.quote_name(Node._meta.db_table),
+                    'name': editor.quote_name('new_parent_inline_fk_idx'),
+                    'using': '',
+                    'columns': editor.quote_name('new_parent_fk_id'),
+                    'extra': '',
+                    'condition': '',
+                },
+            )
+        self.assertIn('new_parent_fk_id', self.get_indexes(Node._meta.db_table))
```

Comment:
Not sure which is preferable but the reason why you couldn't use `add_index` was because the field was missing from the model

```diff
diff --git a/tests/schema/tests.py b/tests/schema/tests.py
index f456e9bb96..d80d98380a 100644
--- a/tests/schema/tests.py
+++ b/tests/schema/tests.py
@@ -292,7 +292,11 @@ class SchemaTests(TransactionTestCase):
         'can_create_inline_fk',
         'allows_multiple_constraints_on_same_fields',
     )
+    @isolate_apps('schema')
     def test_add_inline_fk_index_update_data(self):
+        class Node(Model):
+            class Meta:
+                app_label = 'schema'
         with connection.schema_editor() as editor:
             editor.create_model(Node)
         # Add an inline foreign key, update data, and an index manually in the
@@ -302,17 +306,9 @@ class SchemaTests(TransactionTestCase):
         parent = Node.objects.create()
         with connection.schema_editor() as editor:
             editor.add_field(Node, new_field)
+            Node._meta.add_field(new_field)
             editor.execute('UPDATE schema_node SET new_parent_fk_id = %s;', [parent.pk])
-            editor.execute(
-                editor.sql_create_index % {
-                    'table': editor.quote_name(Node._meta.db_table),
-                    'name': editor.quote_name('new_parent_inline_fk_idx'),
-                    'using': '',
-                    'columns': editor.quote_name('new_parent_fk_id'),
-                    'extra': '',
-                    'condition': '',
-                },
-            )
+            editor.add_index(Node, Index(fields=['new_parent_fk'], name='new_parent_inline_fk_idx'))
         self.assertIn('new_parent_fk_id', self.get_indexes(Node._meta.db_table))

     @skipUnlessDBFeature('supports_foreign_keys')
```

Reply:
Thanks for an idea :1st_place_medal: , I updated tests.